zookeeper-server/src/test/java/org/apache/zookeeper/server/ZooKeeperServerTest.java

```diff
@@ -135,4 +138,32 @@ public void testInvalidSnapshot() {
         }
     }
 
+    @Test
+    public void testClientZxidAhead() {
+        ZooKeeperServer zooKeeperServer = new ZooKeeperServer();
+        final ZKDatabase zkDatabase = new ZKDatabase(mock(FileTxnSnapLog.class));
+        zooKeeperServer.setZKDatabase(zkDatabase);
+
+        final ByteBuffer output = ByteBuffer.allocate(30);
+        // serialize a connReq
+        output.putInt(1);
+        // lastZxid
+        output.putLong(99L);
+        output.putInt(500);
+        output.putLong(123L);
+        output.putInt(1);
+        output.put((byte) 1);
+        output.put((byte) 1);
+        output.flip();
+
+        try {
+            final NIOServerCnxn nioServerCnxn = mock(NIOServerCnxn.class);
+            zooKeeperServer.processConnectRequest(nioServerCnxn, output);
+        } catch (Exception e) {
+            // expect
+            assertTrue(TestServerCnxn.instanceofCloseRequestException(e));
```

Comment:
Why do you need a subclass to access these methods?

Reply:
https://github.com/apache/zookeeper/pull/1518#issuecomment-722369988

It’s necessary to do that. 
'org.apache.zookeeper.server.ServerCnxn.CloseRequestException' has protected access in 'org.apache.zookeeper.server.ServerCnxn', so I can't use it in ReadOnlyZooKeeperServerTest.java.
Do you have a better way to sovle this?

Reply:
If it's necessary for testing, feel free to make it public and annotate with `VisibleForTesting`. We do this every so often, it's neater than adding testing classes all over the place. Thoughts?

Reply:
You could also extract the `CloseRequestException` class in a separate file. It doesn't have to be a nested class.

Reply:
> If it's necessary for testing, feel free to make it public and annotate with `VisibleForTesting`. We do this every so often, it's neater than adding testing classes all over the place. Thoughts?

@anmolnar Subclassing for testing is a valid strategy for testing that requires access to protected methods. The `VisibleForTesting` annotation isn't a great mitigation for unintentional API leakage and misuse, since the compiler won't catch improper uses of it, and neither will most IDEs. Also, it adds an additional dependency on wherever you're getting that annotation imported from (presumably Guava), when you might otherwise not need that dependency. I would encourage the use of subclassing to test protected methods, if it's simple enough to do so (as long as it's reasonable to put the subclass as an inner class inside the test itself). I would resort to increasing the visibility to public, only as a last resort. It is particularly important in ZooKeeper, to limit the public visibility on APIs, because of the fact that ZooKeeper doesn't declare specific packages or classes as "public API", so users of ZooKeeper must rely on the visibility and other factors to infer stability, and they may not notice an annotation that does not trigger any compiler or IDE warning, and incorrectly infer that it is stable when it isn't.

Reply:
I don't agree with subclassing for testing is a good strategy. I saw people hundreds of times reusing each other's subclasses with some modification just because it's already present and convenient to add some things. The result is a complete mess in the testing code, so I highly discourage people using it. Especially introducing it in ZooKeeper at this point in time.

A more preferred way is mocking: that's essentially subclassing and produces much neater code for testing. Maybe not as convenient as adding a new subclass, but it's all about getting the habit of using mocking libraries.

`VisibleForTesting` annotation: I don't have much experience with that, but that must be a reason why projects of Hadoop are using it all over the place. There should be some mechanism on the client side to trigger a warning when somebody is trying to access such member. Either way, ZooKeeper code is heavily using it already, so there's not much harm using is here as well.

In this particular case I think the best would be to extract the exception in a separate file under the same package. I don't see value in nested exception classes anyway.

Reply:
As a tradeoff we can use e.getClass().getName() and make the assertion against the string
this why we don't have to add that trick and also we do not have to add a new public class

@anmolnar  would it work for you ?

Reply:
@eolivelli Yep, sounds like a good compromise.

Reply:
> I don't agree with subclassing for testing is a good strategy.
> A more preferred way is mocking

To be clear, I don't think it's necessarily a *good* strategy, just that in *some* circumstances it might be an appropriate way to test something that you don't otherwise want accessible with higher visibility. I agree mocking is better, whenever possible.

> `VisibleForTesting` annotation: I don't have much experience with that, but that must be a reason why projects of Hadoop are using it all over the place.

I would caution against the idea that popularity is a good measure of reason. :wink:

My 2 cents already having been provided, I don't see any serious problems with any of the discussed strategies. They all seem to have some pros and cons. :smiley_cat:

Reply:
> As a tradeoff we can use e.getClass().getName() and make the assertion against the string
> this why we don't have to add that trick and also we do not have to add a new public class
> 
> @eolivelli It works for "e instanceof NIOServerCnxn.CloseRequestException", but can't know what disconnect reason is.

Reply:
> If it's necessary for testing, feel free to make it public and annotate with `VisibleForTesting`. We do this every so often, it's neater than adding testing classes all over the place. Thoughts?

@anmolnar I have a question, Which package is' VisibleForTesting '?